### PR TITLE
[loader] Check AppContext.BaseDirectory on netcore in preload hook

### PR DESCRIPTION
**Pull request by CoffeeFlux**

### CODE REVIEW

File: mono/metadata/appdomain.c [hunk 1]
```diff
@@ -2444,6 +2470,18 @@ mono_domain_assembly_preload (MonoAssemblyLoadContext *alc,
...
+	char **search_path = g_new0 (char *, 2);
+	search_path [1] = NULL;
+
+	char *base_dir = get_app_context_base_directory (error);
+	search_path [0] = base_dir;
+	mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_ASSEMBLY, "Domain %s (%p) ApplicationBase is %s\n", domain->friendly_name, domain, base_dir);
+
+	result = real_load (search_path, aname->culture, aname->name, &req);
+
+	g_free (search_path);
```
Comment (lambdageek) [inline mono/metadata/appdomain.c]: need to g_free `base_dir`, too.

File: mono/metadata/appdomain.c [hunk 2]
```diff
@@ -2444,6 +2470,18 @@ mono_domain_assembly_preload (MonoAssemblyLoadContext *alc,
 	req.request.predicate = predicate;
 	req.request.predicate_ud = predicate_ud;
 
+#ifdef ENABLE_NETCORE
+	char **search_path = g_new0 (char *, 2);
+	search_path [1] = NULL;
+
+	char *base_dir = get_app_context_base_directory (error);
```
Comment (lambdageek) [inline mono/metadata/appdomain.c]: Probably need a `mono_runtime_get_no_exec()` early exit for AOT somewhere around here.

File: mono/metadata/appdomain.c [hunk 3]
```diff
@@ -2408,6 +2411,27 @@ real_load (gchar **search_path, const gchar *culture, const gchar *name, const M
...
 
+static char *
+get_app_context_base_directory (MonoError *error)
+{
+	MONO_STATIC_POINTER_INIT (MonoMethod, get_basedir)
+
+		ERROR_DECL (local_error);
+		MonoClass *app_context = mono_class_get_app_context_class ();
+		g_assert (app_context);
+		get_basedir = mono_class_get_method_from_name_checked (app_context, "get_BaseDirectory", -1, 0, local_error);
```
Comment (lambdageek) [inline mono/metadata/appdomain.c]: Needs a linker descriptor update.

File: mono/metadata/appdomain.c [hunk 4]
```diff
@@ -2444,6 +2470,18 @@ mono_domain_assembly_preload (MonoAssemblyLoadContext *alc,
 	req.request.predicate = predicate;
 	req.request.predicate_ud = predicate_ud;
 
+#ifdef ENABLE_NETCORE
+	char **search_path = g_new0 (char *, 2);
```
Comment (lambdageek) [inline mono/metadata/appdomain.c]: (minor): `search_path` could be allocated on the stack.

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Pull request merged**